ipm_library/ipm_library/ipm.py

```diff
@@ -0,0 +1,155 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+from ipm_library import utils
+from ipm_library.exceptions import NoIntersectionError
+from ipm_msgs.msg import PlaneStamped
+import numpy as np
+from sensor_msgs.msg import CameraInfo
+from std_msgs.msg import Header
+from tf2_geometry_msgs import PointStamped
+import tf2_ros
+
+
+class IPM:
+    _camera_info: Optional[CameraInfo] = None
+
+    def __init__(
+            self,
+            tf_buffer: tf2_ros.Buffer,
+            camera_info: Optional[CameraInfo] = None) -> None:
+        """
+        Create a new inverse perspective mapper instance.
+
+        :param tf_buffer: This module needs access to a sufficiently large tf2 buffer
+        :param camera_info: `CameraInfo` Message containing the
+            camera intrinsics, camera frame, ...
+            The camera info can be updated later on using the setter or
+            provided directly if it is unlikly to change
+        """
+        # TF needs a listener that is init in the node context, so we need a reference
+        self._tf_buffer = tf_buffer
+        self.set_camera_info(camera_info)
+
+    def set_camera_info(self, camera_info: CameraInfo) -> None:
+        """
+        Set a new `CameraInfo` message.
+
+        :param camera_info: The updated camera info message.
+        """
+        self._camera_info = camera_info
+
+    def camera_info_recived(self) -> bool:
+        """
+        Return if `CameraInfo` message has been recived.
+
+        :returns: If the message was recived
+        """
+        return self._camera_info is not None
+
+    def project_point(
+            self,
+            plane: PlaneStamped,
+            point: PointStamped,
+            output_frame: Optional[str] = None) -> PointStamped:
+        """
+        Projects a `PointStamped` onto a given plane using the latest CameraInfo intrinsics.
+
+        :param plane: Plane in which the projection should happen
+        :param point: Point that should be projected
+        :param output_frame: TF2 frame in which the output should be provided
+        :returns: The point projected onto the given plane in the output frame
+        """
+        # Convert point to numpy and utilize numpy projection function
+        np_point = self.project_points(
+            plane,
+            np.array([[point.point.x, point.point.y, point.point.z]]),
+            point.header,
+            output_frame=None)[0]
```

Comment:
I'm a bit rusty with Python, where can I read up on more about this assignment operator being used as an argument?

Reply:
https://stackoverflow.com/a/1419160

One could just leave it away in this case, but it is often used for optional params as not all of them need to be set, therefore a positional assignment does not make any sense. In most cases, you are able to set the param by name even if it is not optional. This can improve readability as the api is more obvious. In this case the parameter is the only optional one, so it could still be accessed by positional encoding, but it is more readable as `None` does give not real clue regarding the purpose of this param. We could also leave it away entirely as it has `None` as the default value in its signature, but the call is more robust against future API changes if we explicitly specify that.

Reply:
Oh of course.... I've really got to shutdown my cpp brain and bring up my python brain...
Thanks for the explanation!